### PR TITLE
2.1.0: add optional async

### PR DESCRIPTION
**Pull request by junderw**
Using setImmediate to allow for browsers to draw the screen and not freeze up.

This includes breaking changes.

Updated tooling as there were some vulnerabilities and deprecated tools.

### CODE REVIEW

**Comment by junderw** [comment]
... on second thought:

1. This only adds a setImmediate loop once for each p (in BIP38 is = 8)
2. Would probably be easier to use promises on a lower level.
3. That would make supporting sync + async harder (duplicate code)

Perhaps just completely doing away with sync altogether is fine.

All browsers support Promises and async await, so IMO should be fine...

**Comment by junderw** [comment]
@jprichardson bump

This would allow BIP38 to use Async and free up the EventLoop for people during decryption etc.

Most browsers have supported Promises and async await for a while now, so making the default async seems ok.

We could also keep default as sync and make the new Async `"scryptAsync"`.

**Comment by jprichardson** [comment]
@junderw added to the repo and package. @lukechilds please be a 2nd pair of eyes for review.

File: lib/scrypt.js [hunk 1]
```diff
@@ -1,180 +1,26 @@
...
+    B,
+    tickCallback
+  } = checkAndInit(key, salt, N, r, p, dkLen, progressCallback)
 
   for (var i = 0; i < p; i++) {
-    smix(B, i * 128 * r, r, N, V, XY)
+    await smix(B, i * 128 * r, r, N, V, XY, _X, B32, x, tickCallback)
   }
 
   return crypto.pbkdf2Sync(key, B, 1, dkLen, 'sha256')
```
Comment (lukechilds) [inline lib/scrypt.js]: Why not use `pbkdf2` instead of `pbkdf2Sync` here?

Comment (lukechilds) [reply lib/scrypt.js]: Oh, just realised it's only doing one iteration, probably no point then.

**Comment by lukechilds** [comment]
Have you thought about exposing `PROMISE_INTERVAL` as an option users can pass in? Different use cases may have different acceptable perf tradeoffs.

I added a `chunkSize` option when I wrote [`base64-async`](https://github.com/lukechilds/base64-async) to give users control over this. 

It could also be worth showing some benchmark stats between sync/async in the readme. I see you mentioned it's slower in the readme but it would be good to give users an idea of the difference.

Also some prior art in `base64-async` for this if it's helpful:
https://github.com/lukechilds/base64-async#performance

And I'm not really advocating for this, just throwing it out there. But if you wanted to ship this without a breaking API change you could expose it as:

```js
const scrypt = require('scryptsy');

scrypt(key, salt, 16384, 8, 1, 64);
await scrypt.async(key, salt, 16384, 8, 1, 64);
```

and then just release as SemVer minor.

**Comment by junderw** [comment]
- [ ] Allow passing `PROMISE_INTERVAL`
- [ ] Move async to subFunction.
- [ ] Mention an example benchmark.

Yeah, after seeing the perf difference, I'm against async being default.

File: README.md [hunk 2]
```diff
@@ -22,20 +22,94 @@ Installation
...
 
 ```js
-var scrypt = require('scryptsy')
-
-var key = "pleaseletmein"
-var salt = "SodiumChloride"
-var data = scrypt(key, salt, 16384, 8, 1, 64)
-console.log(data.toString('hex'))
-// => 7023bdcb3afd7348461c06cd81fd38ebfda8fbba904f8e3ea9b543f6545da1f2d5432955613f0fcf62d49705242a9af9e61e85dc0d651e40dfcf017b45575887
+const { scrypt, scryptSync } = require('scryptsy')
```
Comment (lukechilds) [inline README.md]:
```suggestion
const scrypt = require('scryptsy')
```

File: README.md [hunk 3]
```diff
@@ -22,20 +22,94 @@ Installation
...
+  var data1 = scryptSync(key, salt, 16384, 8, 1, 64)
+  console.log(data1.toString('hex'))
+  // => 7023bdcb3afd7348461c06cd81fd38ebfda8fbba904f8e3ea9b543f6545da1f2d5432955613f0fcf62d49705242a9af9e61e85dc0d651e40dfcf017b45575887
+
+  // async is actually slower, but it will free up the event loop occasionally
+  // which will allow for front end GUI elements to update and cause it to not
+  // freeze up.
+  // See benchmarks below
+  // Passing 300 below means every 300 iterations internally will call setImmediate once
+  var data2 = await scrypt(key, salt, 16384, 8, 1, 64, undefined, 300)
```
Comment (lukechilds) [inline README.md]:
```suggestion
  var data2 = await scrypt.async(key, salt, 16384, 8, 1, 64, undefined, 300)
```

File: README.md [hunk 4]
```diff
@@ -22,20 +22,94 @@ Installation
...
-var salt = "SodiumChloride"
-var data = scrypt(key, salt, 16384, 8, 1, 64)
-console.log(data.toString('hex'))
-// => 7023bdcb3afd7348461c06cd81fd38ebfda8fbba904f8e3ea9b543f6545da1f2d5432955613f0fcf62d49705242a9af9e61e85dc0d651e40dfcf017b45575887
+const { scrypt, scryptSync } = require('scryptsy')
+
+async function main () {
+  var key = "pleaseletmein"
+  var salt = "SodiumChloride"
+  var data1 = scryptSync(key, salt, 16384, 8, 1, 64)
```
Comment (lukechilds) [inline README.md]:
```suggestion
  var data1 = scrypt(key, salt, 16384, 8, 1, 64)
```

**Comment by lukechilds** [comment]
Also, what about accepting an object as an argument? Maybe that's outside the scope of this PR and it's maybe not worth it if you'd like to avoid making breaking changes.

But I think it would help with both readability of user code and also be a nicer API.

For example, take the code from the readme example:

```js
scrypt.async(key, salt, 16384, 8, 1, 64, undefined, 300)
```

If I encounter that in a codebase it's not very clear what those values mean. Unless someone explicitly declares descriptive variable names for each parameter just to be used once in that function for readability. With an object all args would be explicitly named.

It's also a bit cumbersome if I want to set `promiseInterval` but I don't want a `progressCallback`. I have to pass `undefined` as the 7th argument to be able to set `promiseInterval`. With an object you could just leave out the `progressCallback` property.

So it would become:

```js
scrypt.async({key, salt, N: 16384, r: 8, p: 1, keyLenBytes: 64, promiseInterval: 300})
```

Which enforces better readability and avoids having to juggle undefined arguments.

But like I said, kinda subjective and maybe outside the scope of this PR. It's just with the extra argument added I feel it's more of an issue.

**Comment by junderw** [comment]
I totally agree. I've become more of a fan of the "one object as an arg" paradigm.

But I think it's out of scope... and for now, having sync and async having a differing API stinks.

**Comment by lukechilds** [comment]
I meant for both `scrypt` and `scrypt.async`, so they'd both have the same object API.

But yeah, not a pressing issue right now.

**Comment by jprichardson** [comment]
>  I've become more of a fan of the "one object as an arg" paradigm.

I love this as well. But agreed, out of scope for this PR.

**Pull request merged**